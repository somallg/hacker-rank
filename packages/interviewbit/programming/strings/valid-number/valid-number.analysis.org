* DONE Doc va phan tich yeu cau
  CLOSED: [2018-08-11 Sat 00:12]
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 00:03]--[2018-08-11 Sat 00:12] =>  0:09
  :END:
Cho 1 chuoi ky tu, kiem tra xem chuoi do co phai la so hay khong

V/d:
- "0" => true
- " 0.1 " => true
- "abc" => false
- "1 a" => false
- "2e10" => true

Ta se co 3 kha nang, de chuoi la so:
- Khong chua ky tu nao khac 0-9
- Neu nhu chua e thi ben trai e phai la valid number co the la float, ben phai e cung la valid number so tu nhien
- So float khi co dau cham va ben trai dau cham la so tu nhien, ben phai cung la so tu nhien

* DONE TDD: Setup test case
  CLOSED: [2018-08-11 Sat 00:21]
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 00:12]--[2018-08-11 Sat 00:21] =>  0:09
  :END:
Done

* DONE Dua ra brute force solution
  CLOSED: [2018-08-11 Sat 00:24]
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 00:21]--[2018-08-11 Sat 00:24] =>  0:03
  :END:
Kiem tra lan luot cac ky tu trong chuoi:
- Neu chi co 0-9 tra ve true
- Neu chua '.' thi kiem tra ben trai phai la so tu nhien, ben phai cung la so tu nhien
- Neu chua 'e' thi kiem tra ben trai e la so float, va ben phai la so tu nhien

* DONE Toi uu vao so sanh cac thuat toan
  CLOSED: [2018-08-11 Sat 00:25]
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 00:24]--[2018-08-11 Sat 00:25] =>  0:01
  :END:
Do phuc tap ~O(n)~ nen chua can toi uu hoa hon

* DONE Bat dau code thuat toan thich hop nhat
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 16:00]--[2018-08-11 Sat 16:17] => 0:17
  :END:
Coding

* DONE Test va fix bug neu co
  CLOSED: [2018-08-11 Sat 17:08]
  :LOGBOOK:
  CLOCK: [2018-08-11 Sat 16:17]--[2018-08-11 Sat 17:08] =>  0:51
  :END:

#+BEGIN: clocktable :scope file :maxlevel 2
#+CAPTION: Clock summary at [2018-08-11 Sat 17:09]
| Headline                               |   Time |
|----------------------------------------+--------|
| *Total time*                           | *1:30* |
|----------------------------------------+--------|
| Doc va phan tich yeu cau               |   0:09 |
| TDD: Setup test case                   |   0:09 |
| Dua ra brute force solution            |   0:03 |
| Toi uu vao so sanh cac thuat toan      |   0:01 |
| Bat dau code thuat toan thich hop nhat |   0:17 |
| Test va fix bug neu co                 |   0:51 |
#+END:
