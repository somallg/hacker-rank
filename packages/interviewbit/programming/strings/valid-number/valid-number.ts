/**
 * ValidNumber
 */

function isFloat(s: string): boolean {
  return isPositiveFloat(s) || isNegativeFloat(s);
}

function isPositiveFloat(s: string): boolean {
  if (s === '') {
    return false;
  }

  const dotIndex: number = s.indexOf('.');
  if (dotIndex !== -1) {
    return (
      (isInteger(s.substr(0, dotIndex)) || dotIndex === 0) &&
      isInteger(s.substr(dotIndex + 1))
    );
  }

  return false;
}

function isNegativeFloat(s: string): boolean {
  if (s[0] === '-') {
    return isPositiveFloat(s.substr(1));
  }

  return false;
}

function isExponential(s: string): boolean {
  if (s === '') {
    return false;
  }

  const eIndex: number = s.indexOf('e');
  if (eIndex !== -1) {
    return (
      isFloatOrInteger(s.substr(0, eIndex)) && isInteger(s.substr(eIndex + 1))
    );
  }

  return false;
}

function isFloatOrInteger(s: string): boolean {
  return isFloat(s) || isInteger(s);
}

function isInteger(s: string): boolean {
  return isNegativeInteger(s) || isPositiveInteger(s);
}

function isNegativeInteger(s: string): boolean {
  if (s[0] === '-') {
    return isPositiveInteger(s.substr(1));
  }

  return false;
}

function isPositiveInteger(s: string): boolean {
  if (s === '') {
    return false;
  }

  return s
    .split('')
    .map((d: string) => d.charCodeAt(0))
    .every((c: number) => c >= 48 && c <= 57);
}

function validNumber(s: string): number {
  // tslint:disable-next-line
  s = s.trim();

  return isInteger(s) || isFloat(s) || isExponential(s) ? 1 : 0;
}

export { validNumber };
