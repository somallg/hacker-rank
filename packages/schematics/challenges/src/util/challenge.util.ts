import { strings } from '@angular-devkit/core';

interface SwitchOptions {
  [key: string]: string;
}

function getChallengeName(path: string): string {
  const [, challenge] = path.split('/');

  return challenge;
}

function getFunctionName(challengeName: string, problem: string): string {
  // set default name to camelCase
  const camelizedProblemName = strings.camelize(problem);
  const classifiedProblemName = strings.classify(problem);

  return (
    ({
      codility: 'solution',
      hackerearth: `solve${classifiedProblemName}`,
      hackerrank: `solve${classifiedProblemName}`,
      interviewbit: camelizedProblemName,
      leetcode: camelizedProblemName,
      spoj: camelizedProblemName
    } as SwitchOptions)[challengeName] || camelizedProblemName
  );
}

function getUrl(challengeName: string, problem: string) {
  return (
    ({
      hackerrank: ` @url https://www.hackerrank.com/challenges/${problem}/problem`
    } as SwitchOptions)[challengeName] || strings.classify(problem)
  );
}

export { getChallengeName, getFunctionName, getUrl };
