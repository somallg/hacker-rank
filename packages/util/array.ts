/*
 * Utils function which generate an array of size length
 * Depend on the value of random, if random is true
 * then it generata an array where each element of the array
 * in range [startFrom, ...maxExclusive - 1].
 * If random is false then it return an array [0...0] of size length
 */

export { generateArray, generateStringArray, generateMatrix, zip };

function generateArray(
  length: number,
  maxExclusive: number = length,
  startFrom: number = 0,
  random: boolean = true
): number[] {
  const res: number[] = [];

  for (let i: number = 0; i < length; i = i + 1) {
    // tslint:disable:insecure-random
    res.push(random ? Math.floor(Math.random() * maxExclusive + startFrom) : 0);
  }

  return res;
}

function generateStringArray(length: number): string[] {
  return generateArray(length).map(String);
}

function generateMatrix(
  length: number,
  maxExclusive: number = length
): number[][] {
  const result: number[][] = [];

  for (let i: number = 0; i < length; i = i + 1) {
    result[i] = generateArray(length, maxExclusive);
  }

  return result;
}

function zip<T>(arrayA: T[], arrayB: T[]): T[][] {
  return arrayA.map((a: T, index: number) => [a, arrayB[index]]);
}
