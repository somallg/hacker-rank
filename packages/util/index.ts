export * from './array';
export * from './memorized';
export * from './string';
export * from './unit-test.util';
