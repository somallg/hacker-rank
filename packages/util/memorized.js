function memorized(fn) {
  var cached = {};

  return function(n) {
    if (cached[n] === undefined) {
      cached[n] = fn(n);
    }

    return cached[n];
  };
}

module.exports = {
  memorized
};
