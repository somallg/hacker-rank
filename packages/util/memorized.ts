type Immutable = string | number;

type MemoFn<InputType, OutputType> = (x: InputType) => OutputType;

function id<InputType>(x: InputType): Immutable {
  if (typeof x === 'number') {
    return x;
  } else if (typeof x === 'string') {
    return x;
  }

  return JSON.stringify(x);
}

function memorized<InputType, OutputType>(
  fn: MemoFn<InputType, OutputType>,
  keyFn = id
): MemoFn<InputType, OutputType> {
  const memo = new Map<Immutable, OutputType>();

  return (x: InputType): OutputType => {
    let cachedValue = memo.get(keyFn(x)) as OutputType;
    if (cachedValue === undefined) {
      cachedValue = fn(x);
      memo.set(keyFn(x), cachedValue);
    }

    return cachedValue;
  };
}

export { memorized };
