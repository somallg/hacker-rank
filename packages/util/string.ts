function generateString(length: number, seed: string = 'x'): string {
  let result: string = '';

  while (length > 0) {
    if (length & 1) {
      result = result + seed;
    }

    /* tslint:disable:no-parameter-reassignment */
    length = length >>> 1;
    seed = seed + seed;
  }

  return result;
}

export { generateString };
