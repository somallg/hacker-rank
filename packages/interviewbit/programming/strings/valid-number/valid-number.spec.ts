/**
 * Test specs for valid-number problem
 * ValidNumber
 */

import { generateArray } from '@challenges/util';

import { validNumber } from './valid-number';

describe('Interviewbit - ValidNumber', () => {
  describe('validNumber', () => {
    describe('Example Tests', () => {
      it('should return correctly for Example 1 input', () =>
        expect(validNumber('0')).toMatchSnapshot());

      it('should return correctly for Example 2 input', () =>
        expect(validNumber('0.1')).toMatchSnapshot());

      it('should return correctly for Example 3 input', () =>
        expect(validNumber('abc')).toMatchSnapshot());

      it('should return correctly for Example 4 input', () =>
        expect(validNumber('3.')).toMatchSnapshot());

      it('should return correctly for Example 5 input', () =>
        expect(validNumber('.2')).toMatchSnapshot());
    });

    describe('Correctness Tests', () => {
      it('should return correctly for Sample 1 input', () =>
        expect(validNumber('1 a')).toMatchSnapshot());

      it('should return correctly for Sample 2 input', () =>
        expect(validNumber('2e10')).toMatchSnapshot());

      it('should return correctly for Sample 3 input', () =>
        expect(validNumber('-01.1e-10')).toMatchSnapshot());

      it('should return correctly for Extreme 1 input', () =>
        expect(validNumber('')).toMatchSnapshot());

      it('should return correctly for Extreme 2 input', () =>
        expect(validNumber('   ')).toMatchSnapshot());
    });

    describe('Performance Tests', () => {
      it('should run for Medium 1 input', () =>
        expect(
          validNumber(
            generateArray(100)
              .map(String)
              .join('')
          )
        ).toBeDefined());

      it('should run for Medium 2 input', () =>
        expect(
          validNumber(
            generateArray(1e3)
              .map(String)
              .join('')
          )
        ).toBeDefined());

      it('should run for Large 1 input', () =>
        expect(
          validNumber(
            generateArray(10e3)
              .map(String)
              .join('')
          )
        ).toBeDefined());

      it('should run for Large 2 input', () =>
        expect(
          validNumber(
            generateArray(50e3)
              .map(String)
              .join('')
          )
        ).toBeDefined());

      it('should run for Large 3 input', () =>
        expect(
          validNumber(
            generateArray(100e3)
              .map(String)
              .join('')
          )
        ).toBeDefined());
    });
  });
});
