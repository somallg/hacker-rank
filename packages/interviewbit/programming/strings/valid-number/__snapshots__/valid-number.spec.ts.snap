// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interviewbit - ValidNumber validNumber Correctness Tests should return correctly for Extreme 1 input 1`] = `0`;

exports[`Interviewbit - ValidNumber validNumber Correctness Tests should return correctly for Extreme 2 input 1`] = `0`;

exports[`Interviewbit - ValidNumber validNumber Correctness Tests should return correctly for Sample 1 input 1`] = `0`;

exports[`Interviewbit - ValidNumber validNumber Correctness Tests should return correctly for Sample 2 input 1`] = `1`;

exports[`Interviewbit - ValidNumber validNumber Correctness Tests should return correctly for Sample 3 input 1`] = `1`;

exports[`Interviewbit - ValidNumber validNumber Example Tests should return correctly for Example 1 input 1`] = `1`;

exports[`Interviewbit - ValidNumber validNumber Example Tests should return correctly for Example 2 input 1`] = `1`;

exports[`Interviewbit - ValidNumber validNumber Example Tests should return correctly for Example 3 input 1`] = `0`;

exports[`Interviewbit - ValidNumber validNumber Example Tests should return correctly for Example 4 input 1`] = `0`;

exports[`Interviewbit - ValidNumber validNumber Example Tests should return correctly for Example 5 input 1`] = `1`;
